File: CellsToServers/HierarchicalTilingApp/SumOfSquares/CornacchiaMethod.cs
using System;
using System.Collections.Generic;

namespace HierarchicalTilingApp.SumOfSquares
{
    public class CornacchiaMethod
    {
        private IntTupleEqualityComparer comparer;

        public CornacchiaMethod(IntTupleEqualityComparer comparer)
        {
            this.comparer = comparer;
        }

        public IntTupleEqualityComparer getComparer()
        {
            return comparer;
        }

        /// <summary>
        /// This method return such int tuple where the components of this tuple have ascending order.
        /// </summary>
        /// <param name="num"></param>
        /// <returns></returns>
        public IntTuple[] applyCornacchiaMethod(int num)
        {
            IntTuple[] intTuples;
            if (num == 1)
	        {
                intTuples = new IntTuple[]
                { 
                    new IntTuple() { Tuple = new int[] { 1, 0 } }, 
                    new IntTuple() { Tuple = new int[] { 0, 1 } } 
                };
	        } 
            else
	        {
                List<IntTuple> intTuplesOfFactors = listSumOfSquaresOfFactors(num);
                intTuples = applyFibonacciIdentity(intTuplesOfFactors);
	        }
            return intTuples;
        }

        /// <summary>
        /// This is an another approach. However, this is slower than the other.
        /// </summary>
        /// <param name="limit"></param>
        /// <returns></returns>
        public List<IntTuple>[] applyCornacchiaMethodForRange(int limit)
        {
            List<IntTuple>[] dictOfSolutions = new List<IntTuple>[limit + 1];
            for (int num = 2; num <= limit; num++)
            {
                IntTuple[] currentTuples = applyCornacchiaMethodPrimitiveSolution(num);
                if (dictOfSolutions[num] != null)
                    dictOfSolutions[num].AddRange(currentTuples);
                else
                {
                    dictOfSolutions[num] = new List<IntTuple>(currentTuples);
                }
                addNumSquareForLaterNum(limit, dictOfSolutions, num);
                addNumMultipleForLaterNum(limit, dictOfSolutions, num, currentTuples);
            }
            return dictOfSolutions;
        }

        private void addNumSquareForLaterNum(int limit, List<IntTuple>[] dictOfSolutions, int num)
        { 
            if (num <= Math.Sqrt(limit))
            {
                int laterNum = num * num;
                IntTuple tupleForLaterNum = new IntTuple()
                {
                    Tuple = new int[] { 0, num }
                };
                updateTuplesOfLaterNum(dictOfSolutions, laterNum, tupleForLaterNum);
            }
        }

        private void addNumMultipleForLaterNum(int limit, List<IntTuple>[] dictOfSolutions, int num, 
            IntTuple[] currentTuples)
        {
            foreach (var intTuple in currentTuples)
            {
                int k = 2;
                while (num <= (double)limit / (double)(k * k))
                {
                    int laterNum = num * k * k;
                    IntTuple tupleForLaterNum = new IntTuple()
                    {
                        Tuple = new int[] { intTuple.Tuple[0] * k, intTuple.Tuple[1] * k }
                    };
                    updateTuplesOfLaterNum(dictOfSolutions, laterNum, tupleForLaterNum);
                    k++;
                    laterNum = num * k * k;
                }
            }
        }

        private void updateTuplesOfLaterNum(List<IntTuple>[] dictOfSolutions, int laterNum,
            IntTuple tupleForLaterNum)
        {
            if (dictOfSolutions[laterNum] != null)
                dictOfSolutions[laterNum].Add(tupleForLaterNum);
            else
            {
                dictOfSolutions[laterNum] = new List<IntTuple>();
                dictOfSolutions[laterNum].Add(tupleForLaterNum);
            }
        }

        private IntTuple[] applyCornacchiaMethodPrimitiveSolution(int num)
        {
            List<IntTuple> container = new List<IntTuple>();
            if (num == 2)
            {
                container.Add(new IntTuple { Tuple = new int[] { 1, 1 } });
            }
            else
            {
                for (int t = 1; t <= num / 2; t++)
                {
                    if (t * t < int.MaxValue)
                    {
                        innerApplyCornacchiaMethodPrimitiveSolution(num, container, t);
                    }
                    else
                    {
                        throw new NotSupportedException("Number " + num + " is too large.");
                    }
                }
            }
            return container.ToArray();
        }

        private void innerApplyCornacchiaMethodPrimitiveSolution(int num, List<IntTuple> container, int t)
        {
            int x, y;
            if (t * t % num == num - 1)
            {
                int r1 = num;
                int r2 = t;
                int tmp;
                while (r2 * r2 >= num)
                {
                    tmp = r2;
                    r2 = (r1 % r2);
                    r1 = tmp;
                }
                if (r1 * r1 > num)
                {
                    x = r2;
                    y = r1 % r2;
                    chooseOrderedIntTuple(container, x, y);
                }
            }
        }

        private List<IntTuple> listSumOfSquaresOfFactors(int num)
        {
            List<IntTuple> intTuplesOfFactors = new List<IntTuple>();
            for (int factor = 2; num > 1; factor++)
            {
                if (num % factor == 0)
                {
                    int exponent = 0;
                    while (num % factor == 0)
                    {
                        num /= factor;
                        exponent++;
                        if ((factor % 4 == 1) || (factor == 2))
                        {
                            intTuplesOfFactors.AddRange(applyCornacchiaMethodPrimitiveSolution(factor));
                        }
                    }
                    // If the prime factor is of the form 4k+3 then its exponent should be even
                    // otherwise the num is not expressible as x^2 + y^2 
                    // due to the corollary of Fermat's theorem on sums of two squares
                    if (factor % 4 == 3)
                    {
                        if (exponent % 2 == 0)
                        {
                            intTuplesOfFactors.Add( 
                                new IntTuple { Tuple = new int[] { 0, (int)Math.Pow(factor, exponent / 2) } }
                                );
                        }
                        else
                        {
                            intTuplesOfFactors = new List<IntTuple>();
                            break;
                        }
                    }
                }
            }
            return intTuplesOfFactors;
        }

        private IntTuple[] applyFibonacciIdentity(List<IntTuple> intTuplesOfFactors)
        {
            HashSet<IntTuple> currentIntTuples = new HashSet<IntTuple>(comparer);
            foreach (var intTuple in intTuplesOfFactors)
            {
                currentIntTuples = applyFibonacciIdentity(currentIntTuples, intTuple);
            }
            IntTuple[] result = new IntTuple[currentIntTuples.Count];
            currentIntTuples.CopyTo(result);
            return result;
        }

        private HashSet<IntTuple> applyFibonacciIdentity(HashSet<IntTuple> currentIntTuples, IntTuple newIntTuple)
        {
            HashSet<IntTuple> result = new HashSet<IntTuple>(comparer);
            if (currentIntTuples.Count == 0)
                result.Add(newIntTuple);
            else
            {
                foreach (var currentIntTuple in currentIntTuples)
                {
                    int n1, n2;
                    int a = currentIntTuple.Tuple[0];
                    int b = currentIntTuple.Tuple[1];
                    int c = newIntTuple.Tuple[0];
                    int d = newIntTuple.Tuple[1];
                    n1 = (int)Math.Abs(a * c - b * d);
                    n2 = (int)Math.Abs(a * d + b * c);
                    chooseOrderedIntTuple(result, n1, n2);
                    n1 = (int)Math.Abs(a * c + b * d);
                    n2 = (int)Math.Abs(a * d - b * c);
                    chooseOrderedIntTuple(result, n1, n2);
                }
            }
            return result;
        }

        private void chooseOrderedIntTuple(ICollection<IntTuple> result, int n1, int n2)
        {
            if (n1 < n2)
            {
                result.Add(new IntTuple { Tuple = new int[] { n1, n2 } });
            }
            else
            {
                result.Add(new IntTuple { Tuple = new int[] { n2, n1 } });
            }
        }
    }
}

File: CellsToServersCppLinux/HierarchicalTilingAppCpp/SumOfSquares/CornacchiaMethod.cpp
#include "CornacchiaMethod.hpp"



namespace SumOfSquares
{
	Vector_t& CornacchiaMethod::applyCornacchiaMethod(int num)
	{
		Vector_t * intTuples = new Vector_t();
		if (num == 1)
		{
			IntTuple * intTuple1 = new IntTuple(2, tuple_1_0);
			intTuples->push_back(intTuple1);
			IntTuple * intTuple2 = new IntTuple(2, tuple_0_1);
			intTuples->push_back(intTuple2);
		} 
		else
		{
			List_t intTuplesOfFactors = listSumOfSquaresOfFactors(num);
			intTuples = applyFibonacciIdentity(intTuplesOfFactors);
		}
		return *intTuples;
	}

	List_t& CornacchiaMethod::listSumOfSquaresOfFactors(int num)
	{
		List_t * intTuplesOfFactors = new List_t();
		for (int factor = 2; num > 1; factor++)
		{
			if (num % factor == 0)
			{
				int exponent = innerListSumOfSquaresOfFactors(intTuplesOfFactors, factor, num);
				// If the prime factor is of the form 4k+3 then its exponent should be even
				// otherwise the num is not expressible as x^2 + y^2 
				// due to the corollary of Fermat's theorem on sums of two squares
				if (factor % 4 == 3)
				{
					if (exponent % 2 == 0)
					{
						int newTuple[2] = { 0, (int)pow(factor, exponent / 2) };
						IntTuple * newIntTuple = new IntTuple(2, newTuple);
						intTuplesOfFactors->push_back(newIntTuple);
					}
					else
					{
						intTuplesOfFactors = new List_t();
						break;
					}
				}
			}
		}
		return *intTuplesOfFactors;
	}

	int CornacchiaMethod::innerListSumOfSquaresOfFactors(List_t * intTuplesOfFactors, int factor,
		int& num)
	{
		int exponent = 0;
		while (num % factor == 0)
		{
			num /= factor;
			exponent++;
			if ((factor % 4 == 1) || (factor == 2))
			{
				List_t primitiveSolutions = applyCornacchiaMethodPrimitiveSolution(factor);
				for(List_t::iterator itr = primitiveSolutions.begin(); 
					itr != primitiveSolutions.end(); itr++)
				{
					intTuplesOfFactors->push_back(*itr);
				}
			}
		}
		return exponent;
	}

	List_t& CornacchiaMethod::applyCornacchiaMethodPrimitiveSolution(int num)
	{
		List_t * container = new List_t();
		if (num == 2)
		{
			IntTuple * newIntTuple = new IntTuple(2, tuple_1_1);
			container->push_back(newIntTuple);
		}
		else
		{
			for (int t = 1; t <= num / 2; t++)
			{
				if (t * t < std::numeric_limits<int>::max())
				{
					innerApplyCornacchiaMethodPrimitiveSolution(num, container, t);
				}
				else
				{
					std::string errmsg("Number ");
					errmsg.append("" + num);
					errmsg.append(" is too large.\n");
					throw std::runtime_error(errmsg);
				}
			}
		}
		return *container;
	}

	void CornacchiaMethod::innerApplyCornacchiaMethodPrimitiveSolution(int num, List_t * container, int t)
	{
		int x, y;
		if (t * t % num == num - 1)
		{
			int r1 = num;
			int r2 = t;
			int tmp;
			while (r2 * r2 >= num)
			{
				tmp = r2;
				r2 = (r1 % r2);
				r1 = tmp;
			}
			if (r1 * r1 > num)
			{
				x = r2;
				y = r1 % r2;
				chooseOrderedIntTuple(container, x, y);
			}
		}
	}

	Vector_t * CornacchiaMethod::applyFibonacciIdentity(List_t& intTuplesOfFactors)
	{
		Uset_t currentIntTuples;
		for(List_t::iterator itr = intTuplesOfFactors.begin(); itr != intTuplesOfFactors.end(); itr++)
		{
			IntTuple * intTuple = *itr;
			currentIntTuples = applyFibonacciIdentity(currentIntTuples, intTuple);
		}
		Vector_t * result = new Vector_t();
		for(Uset_t::iterator itr = currentIntTuples.begin(); itr != currentIntTuples.end(); itr++)
		{
			result->push_back(*itr);
		}
		return result;
	}

	Uset_t& CornacchiaMethod::applyFibonacciIdentity(Uset_t& currentIntTuples, IntTuple * newIntTuple)
	{
		Uset_t * result = new Uset_t();
		if (currentIntTuples.size() == 0)
			result->insert(newIntTuple);
		else
		{
			for(Uset_t::iterator itr = currentIntTuples.begin(); itr != currentIntTuples.end(); itr++)
			{
				IntTuple * currentIntTuple = *itr;
				int n1, n2;
				int a = currentIntTuple->getTuple()[0];
				int b = currentIntTuple->getTuple()[1];
				int c = newIntTuple->getTuple()[0];
				int d = newIntTuple->getTuple()[1];
				n1 = abs(a * c - b * d);
				n2 = abs(a * d + b * c);
				chooseOrderedIntTuple(result, n1, n2);
				n1 = abs(a * c + b * d);
				n2 = abs(a * d - b * c);
				chooseOrderedIntTuple(result, n1, n2);
			}
		}
		return *result;
	}
	
	void CornacchiaMethod::chooseOrderedIntTuple(List_t * result, int n1, int n2)
	{
		if (n1 < n2)
		{
			int * newTuple = new int[2]; newTuple[0] = n1; newTuple[1] = n2;
			IntTuple * newIntTuple = new IntTuple(2, newTuple);
			result->push_back(newIntTuple);
		}
		else
		{
			int * newTuple = new int[2]; newTuple[0] = n2; newTuple[1] = n1;
			IntTuple * newIntTuple = new IntTuple(2, newTuple);
			result->push_back(newIntTuple);
		}
	}
	
	void CornacchiaMethod::chooseOrderedIntTuple(Uset_t * result, int n1, int n2)
	{
		if (n1 < n2)
		{
			int * newTuple = new int[2]; newTuple[0] = n1; newTuple[1] = n2;
			IntTuple * newIntTuple = new IntTuple(2, newTuple);
			result->insert(newIntTuple);
		}
		else
		{
			int * newTuple = new int[2]; newTuple[0] = n2; newTuple[1] = n1;
			IntTuple * newIntTuple = new IntTuple(2, newTuple);
			result->insert(newIntTuple);
		}
	}
}